### DISCUSSION TITLE
GreenField: Wallet API

### DISCUSSION THREAD

**Discussion by Kukks** [btcpayserver/btcpayserver#2143] [Technical Support]
Currently, a `WalletId` consists of a `store id` and a `crypto code` combined together with the format of `S-{storeId}-{cryptoCode}`.  A wallet is 1-1 with a store (that is what the `S-` prefix signals) and one of its on-chain payment methods. In the future, we would like to allow the creation of wallets which are not bound to a store and I would also like to extend the wallet concept to other payment methods that are not only on-chain, such as Lightning. 

I believe that `cryptoCode` should be removed from the id (wallet ids are not saved in the DB, only computed based on store and crypto in the url) and instead have the wallet crypto type under a nested url route (instead of ``S-gdhasdhsdgh-BTC`  it would be `S-gdhasdhsdgh/BTC`). 

Further to that, I also propose that we move from `crypto code` to `Payment Method Id`, which consists of `Crypto code` and `Payment type`. This allows us to have a collection of different payment method wallets under one Id, in an exact construction to a store (a store has many payment methods of many currencies) . 

API endpoint considerations:

```
GET api/v1/wallets - Get list of wallets
GET api/v1/wallets/{walletId} - Get list of configured methods of a wallet and general wallet settings
GET api/v1/wallets/{walletId}/{paymentMethodId} - Get generic details of specific configured wallet method

//configuring a wallet requires you use the more specific endpoints
GET api/v1/wallets/{walletId}/{paymentType}/{cryptoCode} - Get overall details of specific configured wallet
PUT api/v1/wallets/{walletId}/{paymentType}/{cryptoCode} - Configure specific wallet
DELETE  api/v1/wallets/{walletId}/{paymentType}/{cryptoCode} - Remove specific wallet

//in reality, a wallet is really an abstraction of a store. When you configure a wallet that has an id that starts with `S`, you are really configuring a store! The below are aliases to the previous configurarion endpoints. 
GET api/v1/stores/{storeId}/payment-methods/{paymentType}/{cryptoCode} - Get overall details of specific configured wallet
PUT api/v1/stores/{storeId}/payment-methods/{paymentType}/{cryptoCode} - Configure specific wallet
DELETE  api/v1/stores/{storeId}/payment-methods/{paymentType}/{cryptoCode} - Remove specific wallet
```

As you can see, we can hit 2 birds with one stone, as the wallet API becomes also a Store Payment Methods API!

For each payment type, we will need separate endpoints as they configuration is different. We will implement only the APIs for `BitcoinLike` and for `LightningLike`, and the altcoin communities are free to implement their own endpoints to manage those payment methods.

The PR https://github.com/btcpayserver/btcpayserver/pull/1767 is almost perfect for the above proposal. 

# Using the actual wallet

Constructing transactions is a whole other beast.  For BitcoinLike, it is as follows:
```
GET api/v1/wallets/{walletId}/{paymentType}/{cryptoCode}/transactions/{txId} - Get a specific tx
GET api/v1/wallets/{walletId}/{paymentType}/{cryptoCode}/transactions - Get list of txs
POST api/v1/wallets/{walletId}/{paymentType}/{cryptoCode}/transactions - Create a tx
POST api/v1/wallets/{walletId}/{paymentType}/{cryptoCode}/psbt - Create a tx

```
We should focus on a MVP for transaction management. For LightningLike, we could most likely mirror parts of the Store LightningNode API.

**Comment by pavlenex** [comment]
@NicolasDorier do you have any guidance on what you'd like to have for the wallet API?

**Comment by NicolasDorier** [comment]
1. I think the API should not be generic. Payment methods are fundamentally different in term of API, wallet APIs will be fundamentally different depending on the type.


2. Given that right now, only store wallet are supported, we should not spend time discussing about a wallet API unrelated to store. This is a waste of time.

So I would say: 

```
GET api/v1/stores/{storeId}/wallets/{paymentType}/{cryptoCode} - Get overall details of specific configured wallet
PUT api/v1/stores/{storeId}/wallets/{paymentType}/{cryptoCode} - Configure specific wallet
```

Seems good, and we should focus only on this for the first release.
Given the APIs can't be generic, this is not a single route, but several route, eg something like two different route:

```
GET api/v1/stores/{storeId}/wallets/OnChain/{cryptoCode} - Get overall details of specific on chain wallet
GET api/v1/stores/{storeId}/wallets/Lightning/{cryptoCode} - Get overall details of specific off chain wallet
```

This is because those would have different APIs, so it can't be generic.

About `api/v1/wallets` I disagree. A walletId has always been `S-WalletId-BTC`, trying to change the meaning will lead to bug and be confusing as hell. But I would say, we can keep this debate for another time. if at least we agree on the  `api/v1/stores/{storeId}/wallets/OnChain/{cryptoCode}` route, we can move on for a while.

If you start working on this, please keep it small. Like only starting by the OnChain wallets, instead of making a huge PR with everything in the kitchen sink.

EDIT:

About `api/v1/wallets`, actually I don't know, I see bad points of keeping S-Walletid-BTC as well, as you said. I would prefer we defer the discussion and focus on what we agree.

**Reply by Kukks** [reply]
> 1. I think the API should not be generic. Payment methods are fundamentally different in term of API, wallet APIs will be fundamentally different depending on the type.

 This is true, I was only refering to maybe fetching any very high-level common data that each would have. Let's leave this out for now as it may not be helpful short-term anyway.

> 2. Given that right now, only store wallet are supported, we should not spend time discussing about a wallet API unrelated to store. This is a waste of time.
> 

My goal is not to develop the decoupled wallet system, but to have an upgrade path that is not painful.  What I wanted to outline is that 
`api/v1/wallets/{walletId}/{paymentType}/{cryptoCode}` is equivalent to `api/v1/store/{storeId}/payment-methods/{paymentType}/{cryptoCode}`

> Seems good, and we should focus only on this for the first release.
> Given the APIs can't be generic, this is not a single route, but several route, eg something like two different route:

Yes, I agree. I only made them appear as routes for the sake of representing all types but internally in code it would be specific controllers for reach Payment type.  Please see https://github.com/btcpayserver/btcpayserver/pull/1767 as an example

> If you start working on this, please keep it small. Like only starting by the OnChain wallets, instead of making a huge PR with everything in the kitchen sink.

Yup, I will resurrect #1767 which focuses only on BitcoinLike. 

> About `api/v1/wallets`, actually I don't know, I see bad points of keeping S-Walletid-BTC as well, as you said. I would prefer we defer the discussion and focus on what we agree.

Sure, let's focus on configuring the payment method as a store first. We should however tackle this before we start the wallet API for constructing transactions to be sure we do not introduce breaking changes in url routes later.
